lib/time.c

```diff
@@ -202,12 +202,61 @@ grn_timeval2str(grn_ctx *ctx, grn_timeval *tv, char *buf, size_t buf_size)
   return ctx->rc;
 }
 
+static grn_rc
+grn_str2timeval_offset_sec(const char *start, const char *end, bool *have_offset, int32_t *gm_offset_sec)
+{
+  const char *position = start;
+  int32_t hours, minutes = 0, sign = 1;
+
+  *have_offset = false;
+  *gm_offset_sec = 0;
+
+  if (position >= end) { return GRN_SUCCESS; }
+
+  switch (*position) {
+  case '+' :
+    sign = 1;
+    *have_offset = true;
+    break;
+  case '-' :
+    sign = -1;
+    *have_offset = true;
+    break;
+  case 'Z' :
+  case 'z' :
+    *have_offset = true;
+    return GRN_SUCCESS;
+  default :
+    return GRN_INVALID_ARGUMENT;
+  }
+
+  position++;
+  if (position >= end) { return GRN_INVALID_ARGUMENT; }
+
+  hours = grn_atoi(position, end, &position);
+  if (hours < 0 || hours >= 24) { return GRN_INVALID_ARGUMENT; }
+
+  if (*position == ':') {
+    position++;
+    if(position == end) { return GRN_INVALID_ARGUMENT; }
```

Comment:
`*end` can be `:`, so if checking `position` with `if(position >= end)`, this line returns `GRN_INVALID_ARGUMENT` even the argument is valid in some cases.
I mean the cases like below.
(This was occurred in `load/time/cast/string/iso8601/valid.test` in my environment)

input value
`2000-01-01T11:01:00.123+01:01`

actual memory
```
start                 end
*                     *
2000-01-01T11:01:00+01:
```

Reply:
Ah, sorry. I wanted to say it for `if (*position == ':') {` not `if (position == end)`.

Reply:
Ah, I see. 
I have moved `position >= end` check into `if (*position == ':')`  for reducing nests.